### DISCUSSION TITLE
[API] api/data response shaping

### DISCUSSION THREAD

**Discussion by MaceWindu** [datalust/seq-tickets#1999] [Help]
Are there any parameters for `api/data` endpoint or a separate enpoint to return data as a list of objects?

We use `api/data` to query for specific event properties using seq `query` but there is one issue with current response - data returned as array of arrays (`Rows[][]` response property), where each record is an array with property values of different types.

I can see how this format is useful for SEQ UI as it minimize payload size, but it is not convenient for us due to limitations of software we use to read API response. It could process array of objects, but not array of arrays.

We currently use our custom middleware to convert data from one format to another but it would be nice if we can read data from SEQ directly in desired format.

**Comment by nblumhardt** [comment]
Hi! At the moment, no - only the arrays-of-arrays format and CSV are implemented (in case CSV is a better option for you?).

If you're able to supply an example ("here's the response we'd get today, here's the response we'd like") I'm keen to improve this, though it'll need to wait for our next feature cycle (we're in the middle of building tracing for Seq 2024.1, currently).

**Reply by MaceWindu** [reply]
E.g. we have API call:
```
/api/data?q=<seq query here>&signal=<signal-id>...
```


Response is:
```json
{
    // list of column names
    "Columns": ["column1", "column2", "column2"],
    "Rows": [
        // event (query result data), each array entry is an array of column values
        [ value-for-column1, value-for-column2, value-for-column3], // record 1
        [ value-for-column1, value-for-column2, value-for-column3], // record 2
        [ value-for-column1, value-for-column2, value-for-column3], // record 3
        ...
    ]
   
    "Statistics": api call statistics, not important here
}
```

What would we like to see:
```json
{
    "Data": [
    // each record is an object (hashtable) with column name as key
    {  // record 1
        "column1": value-of-column1,
        "column2": value-of-column2,
        "column3": value-of-column3
    },
    {  // record 2
        "column1": value-of-column1,
        "column2": value-of-column2,
        "column3": value-of-column3
    },
    ....
    ]
    // other information like statistics is not important for us, but could also present
}
```

Actually regarding `/api/data` API we would like to see another improvement. Currently it accepts query text because this is how seq UI works - user types query and UI requests data using query text.
For our use case we already have predefined query, stored in SEQ, so it makes sense to pass query id as we don't want to hardcode query text in client application. How we solve it right now:
- use query API to get query text by query id
- pass query text to `api/data`

**Reply by nblumhardt** [reply]
Thanks for sharing your scenario, we'll give this all some thought 👍
